File: archivist/src/test/java/com/zorroa/archivist/service/AssetServiceTests.java
package com.zorroa.archivist.service;

import com.google.common.collect.ImmutableList;
import com.zorroa.archivist.AbstractTest;
import com.zorroa.archivist.domain.*;
import com.zorroa.archivist.sdk.security.Groups;
import com.zorroa.sdk.domain.Document;
import com.zorroa.sdk.domain.PagedList;
import com.zorroa.sdk.domain.Pager;
import com.zorroa.sdk.processor.Source;
import com.zorroa.sdk.schema.PermissionSchema;
import com.zorroa.sdk.search.AssetSearch;
import com.zorroa.sdk.util.Json;
import org.junit.Before;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.nio.file.Paths;

import static org.junit.Assert.*;

/**
 * Created by chambers on 9/1/16.
 */
public class AssetServiceTests extends AbstractTest {

    @Autowired
    CommandService commandService;

    @Before
    public void init() {
        addTestAssets("set04/standard");
        refreshIndex();
    }

    @Test
    public void testGetAsset() {
        PagedList<Document> assets = assetService.getAll(Pager.first());
        for (Document a: assets) {
            assertEquals(a.getId(),
                    assetService.get(Paths.get(a.getAttr("source.path", String.class))).getId());
        }
    }

    @Test
    public void tetDelete() {
        PagedList<Document> assets = assetService.getAll(Pager.first());
        for (Document a: assets) {
            assertTrue(assetService.delete(a.getId()));
        }
    }

    @Test
    public void testIndexWithLink() throws InterruptedException {
        Source builder = new Source(getTestImagePath("set01/toucan.jpg"));
        builder.addToLinks("foo", "abc123");

        Document asset1 = assetService.index(builder);
        assertEquals(ImmutableList.of(1),
                asset1.getAttr("zorroa.links.foo"));
    }

    @Test
    public void testIndexCheckOrigin() throws InterruptedException {
        Source builder = new Source(getTestImagePath("set01/toucan.jpg"));
        Document asset1 = assetService.index(builder);

        logger.info("{}", Json.prettyString(asset1.getDocument()));

        assertNotNull(asset1.getAttr("zorroa.timeCreated"));
        assertNotNull(asset1.getAttr("zorroa.timeModified"));
        assertEquals(asset1.getAttr("zorroa.timeCreated", String.class),
                asset1.getAttr("zorroa.timeModified", String.class));

        refreshIndex(1000);
        Source builder2 = new Source(getTestImagePath("set01/toucan.jpg"));
        Document asset2 = assetService.index(builder2);
        assertNotEquals(asset2.getAttr("zorroa.timeCreated", String.class),
                asset2.getAttr("zorroa.timeModified", String.class));
    }

    @Test
    public void testIndexWithPermission() throws InterruptedException {
        Permission p = permissionService.getPermission(Groups.EVERYONE);

        Source builder = new Source(getTestImagePath("set01/toucan.jpg"));
        builder.addToPermissions(Groups.EVERYONE, 7);

        Document asset1 = assetService.index(builder);
        assertEquals(ImmutableList.of(p.getId().toString()),
                asset1.getAttr("zorroa.permissions.read"));

        assertEquals(ImmutableList.of(p.getId().toString()),
                asset1.getAttr("zorroa.permissions.write"));

        assertEquals(ImmutableList.of(p.getId().toString()),
                asset1.getAttr("zorroa.permissions.export"));
    }

    @Test
    public void testIndexWithReadOnlyPermission() throws InterruptedException {
        Permission p = permissionService.getPermission(Groups.EVERYONE);

        Source builder = new Source(getTestImagePath("set01/toucan.jpg"));
        builder.addToPermissions(Groups.EVERYONE, 1);

        Document asset1 = assetService.index(builder);
        assertEquals(ImmutableList.of(p.getId().toString()),
                asset1.getAttr("zorroa.permissions.read"));

        assertNotEquals(ImmutableList.of(p.getId().toString()),
                asset1.getAttr("zorroa.permissions.write"));

        assertNotEquals(ImmutableList.of(p.getId().toString()),
                asset1.getAttr("zorroa..export"));
    }

    @Test
    public void testIndexRemovePermissions() throws InterruptedException {
        Permission p = permissionService.getPermission(Groups.EVERYONE);

        Source builder = new Source(getTestImagePath("set01/toucan.jpg"));
        builder.addToPermissions(Groups.EVERYONE, 7);
        Document asset1 = assetService.index(builder);
        refreshIndex();

        Source builder2 = new Source(getTestImagePath("set01/toucan.jpg"));
        builder.addToPermissions(Groups.EVERYONE, 1);
        Document asset2 = assetService.index(builder);

        // Should only end up with read.
        assertEquals(ImmutableList.of(p.getId().toString()),
                asset2.getAttr("zorroa.permissions.read"));

        assertNotEquals(ImmutableList.of(p.getId().toString()),
                asset2.getAttr("zorroa.permissions.write"));

        assertNotEquals(ImmutableList.of(p.getId().toString()),
                asset2.getAttr("zorroa.permissions.export"));

        Document asset3 = assetService.get(asset2.getId());

        // Should only end up with read.
        assertEquals(ImmutableList.of(p.getId().toString()),
                asset3.getAttr("zorroa.permissions.read"));

        assertNotEquals(ImmutableList.of(p.getId().toString()),
                asset3.getAttr("zorroa.permissions.write"));

        assertNotEquals(ImmutableList.of(p.getId().toString()),
                asset3.getAttr("zorroa.permissions.export"));
    }

    @Test
    public void testSetPermissions() throws InterruptedException {

        Permission p = permissionService.getPermission("user::user");
        Acl acl = new Acl();
        acl.add(new AclEntry(p.getId(), Access.Read));

        CommandSpec spec = new CommandSpec();
        spec.setType(CommandType.UpdateAssetPermissions);
        spec.setArgs(new Object[] {
                new AssetSearch(),
                acl
        });

        Command cmd = commandService.submit(spec);
        commandService.run(commandService.refresh(cmd));

        for(;;) {
            Thread.sleep(200);
            cmd = commandService.refresh(cmd);
            if (cmd.getState().equals(JobState.Finished)) {
                logger.info("Command {} finished", cmd);
                break;
            }
        }
        refreshIndex();

        PagedList<Document> assets = assetService.getAll(Pager.first());
        assertEquals(2, assets.size());

        PermissionSchema schema = assets.get(0).getAttr("zorroa.permissions", PermissionSchema.class);
        assertTrue(schema.getRead().contains(p.getId()));
        assertFalse(schema.getWrite().contains(p.getId()));
        assertFalse(schema.getExport().contains(p.getId()));

        schema = assets.get(1).getAttr("zorroa.permissions", PermissionSchema.class);
        assertTrue(schema.getRead().contains(p.getId()));
        assertFalse(schema.getWrite().contains(p.getId()));
        assertFalse(schema.getExport().contains(p.getId()));

    }
}

File: archivist/src/test/kotlin/com/zorroa/archivist/service/AssetServiceTests.kt
package com.zorroa.archivist.service

import com.zorroa.archivist.AbstractTest
import com.zorroa.archivist.domain.*
import com.zorroa.archivist.repository.AuditLogDao
import com.zorroa.archivist.search.AssetSearch
import com.zorroa.common.schema.PermissionSchema
import com.zorroa.security.Groups
import org.junit.Before
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import java.util.*
import kotlin.test.*

class AssetServiceTests : AbstractTest() {

    @Autowired
    lateinit var auditLogDao: AuditLogDao

    override fun requiresElasticSearch() : Boolean {
        return true
    }

    @Before
    fun init() {
        addTestAssets("set04/standard")
    }

    @Test
    fun testCreateWithWatchedField() {
        System.setProperty("archivist.auditlog.watched-fields", "foo")
        try {
            val assets =  searchService.search(Pager.first(), AssetSearch())
            assets[0].setAttr("foo", "bar")
            assetService.batchCreateOrReplace(BatchCreateAssetsRequest(assets.list))
            val change = jdbc.queryForMap(
                    "SELECT * FROM auditlog WHERE pk_asset=?::uuid AND int_type=?",
                    assets[0].id, AuditLogType.Changed.ordinal)
            assertEquals(assets[0].id, change["pk_asset"].toString())
            assertEquals(AuditLogType.Changed.ordinal, change["int_type"] as Int)

        } finally {
            System.clearProperty("archivist.auditlog.watched-fields")
        }
    }

    @Test
    fun testBatchCreateOrReplace() {
        var assets = searchService.search(Pager.first(), AssetSearch())
        for (asset in assets) {
            validateSystemAttrsExist(asset)
            assertEquals(asset.getAttr("system.timeCreated",
                    String::class.java),
                    asset.getAttr("system.timeModified", String::class.java))
        }

        Thread.sleep(1000)
        val rsp = assetService.batchCreateOrReplace(BatchCreateAssetsRequest(assets.list))
        assertEquals(2, rsp.replacedAssetIds.size)
        assertEquals(0, rsp.createdAssetIds.size)

        refreshIndex()
        assets = searchService.search(Pager.first(), AssetSearch())
        for (asset in assets) {
            validateSystemAttrsExist(asset)
            assertNotEquals(asset.getAttr("system.timeCreated",
                    String::class.java),
                    asset.getAttr("system.timeModified", String::class.java))
        }
    }

    @Test
    fun testBatchCreateOrReplaceWithLinks() {
        var assets = searchService.search(Pager.first(), AssetSearch())
        assetService.batchCreateOrReplace(BatchCreateAssetsRequest(assets.list))

        val ids = assets.map { it.id }
        val folderId = UUID.randomUUID()
        assertEquals(2, assetService.addLinks(
                LinkType.Folder, folderId, BatchUpdateAssetLinks(ids)).updatedAssetIds.size)


        assetService.batchCreateOrReplace(BatchCreateAssetsRequest(assets.list))
        assets = searchService.search(Pager.first(), AssetSearch())
        assertTrue(assets.size() > 0)
        assets.list.forEach {
            assertTrue(it.attrExists("system.links"))
        }
    }

    @Test
    fun testBatchCreateOrReplaceWithManualEdits() {

        val field = fieldSystemService.getField("media.title")

        var assets = searchService.search(Pager.first(), AssetSearch())
        assets.list.forEach {
            assetService.createFieldEdit(FieldEditSpec(UUID.fromString(it.id), field.id, null,"bilbo"))
        }

        assetService.batchCreateOrReplace(BatchCreateAssetsRequest(assets.list))
        assets = searchService.search(Pager.first(), AssetSearch())
        assertTrue(assets.size() > 0)
        assets.list.forEach {
            assertEquals("bilbo", it.getAttr("media.title", String::class.java))
        }
    }

    @Test
    fun testCreateOrReplace() {
        val asset = searchService.search(Pager.first(), AssetSearch())[0]
        asset.setAttr("foo", "bar")
        val newAsset = assetService.createOrReplace(asset)
        assertEquals(newAsset.getAttr("foo", String::class.java), "bar")
    }

    @Test
    fun testGet() {
        val asset1 = searchService.search(Pager.first(), AssetSearch())[0]
        val asset2 = assetService.get(asset1.id)
        assertEquals(asset1.id, asset2.id)
        assertEquals(asset1.getAttr("source.path", String::class.java),
                asset2.getAttr("source.path", String::class.java))
    }

    @Test
    fun testBatchUpdate() {

        var updates = mutableMapOf<String, UpdateAssetRequest>()
        searchService.search(Pager.first(), AssetSearch()).forEach {doc ->
            updates[doc.id] = UpdateAssetRequest(mapOf("foos" to "ball"))
        }
        val rsp = assetService.batchUpdate(BatchUpdateAssetsRequest(updates))
        assertEquals(2, rsp.updatedAssetIds.size)
        refreshIndex()

        val search =  searchService.search(Pager.first(), AssetSearch())
        assertEquals(2, search.size())
        for (doc: Document in search.list) {
            assertEquals("ball", doc.getAttr("foos", String::class.java))
        }
    }

    @Test
    fun testBatchUpdateRemove() {
        val batch = mutableMapOf<String, Map<String, Any?>>()

        var updates= mutableMapOf<String, UpdateAssetRequest>()
        searchService.search(Pager.first(), AssetSearch()).forEach {doc ->
            updates[doc.id] = UpdateAssetRequest(null, listOf("source"))
        }
        val rsp = assetService.batchUpdate(BatchUpdateAssetsRequest(updates))
        assertEquals(2, rsp.updatedAssetIds.size)
        refreshIndex()

        val search =  searchService.search(Pager.first(), AssetSearch())
        assertEquals(2, search.size())
        for (doc: Document in search.list) {
            assertNull(doc.getAttr("source"))
        }
    }

    @Test
    fun testUpdate() {
        val asset1 = searchService.search(Pager.first(), AssetSearch())[0]
        val asset2 = assetService.update(asset1.id, mapOf("foo" to "bar"))
        assertEquals(asset1.id, asset2.id)
        assertEquals(asset1.getAttr("source.path", String::class.java),
                asset2.getAttr("source.path", String::class.java))
        assertEquals(asset2.getAttr("foo", String::class.java), "bar")
    }

    @Test
    fun testUpdateWithUpdateAssetRequest() {
        val asset1 = searchService.search(Pager.first(), AssetSearch())[0]
        val updated = assetService.update(asset1.id, UpdateAssetRequest(
                mapOf("foo" to "bar"),
                listOf("source.fileSize"))
        )
        assertTrue(updated.isSuccess())
        val asset2 = assetService.get(asset1.id)
        assertEquals(asset2.getAttr("foo", String::class.java), "bar")
        assertNull(asset2.getAttr("source.fileSize"))
    }

    @Test
    fun testDelete() {

        val page = searchService.search(Pager.first(), AssetSearch())
        assertTrue(assetService.delete(page[0].id))
        refreshIndex()

        val leftOvers = searchService.search(Pager.first(), AssetSearch())
        assertEquals(page.size() - leftOvers.size(), 1)
    }

    @Test
    fun testBatchDelete() {
        val page = searchService.search(Pager.first(), AssetSearch())
        val ids = page.map { it.id }
        val rsp = assetService.batchDelete(ids)
        assertEquals(0, rsp.errors.size)
        assertEquals(2, rsp.deletedAssetIds.size)
        assertEquals(2, rsp.totalRequested)
    }

    @Test
    fun testAddLinks() {
        val page = searchService.search(Pager.first(), AssetSearch())
        val ids = page.map { it.id }
        val folderId = UUID.randomUUID()
        assertEquals(2, assetService.addLinks(
                LinkType.Folder, folderId, BatchUpdateAssetLinks(ids)).updatedAssetIds.size)
        assertEquals(0, assetService.addLinks(
                LinkType.Folder, folderId, BatchUpdateAssetLinks(ids)).erroredAssetIds.size)
    }

    @Test
    fun testAddChildLinks() {
        val page = searchService.search(Pager.first(), AssetSearch())
        assetService.update(page[0].id, mapOf("media.clip.parent" to page[1].id))
        refreshIndex()

        val folderId = UUID.randomUUID()
        assertEquals(1,
                assetService.addLinks(LinkType.Folder, folderId,
                        BatchUpdateAssetLinks(null, listOf(page[1].id), AssetSearch())).updatedAssetIds.size)
    }

    @Test
    fun testRemoveLinks() {
        val page = searchService.search(Pager.first(), AssetSearch())
        val ids = page.map { it.id }
        val folderId = UUID.randomUUID()
        assertEquals(2, assetService.addLinks(
                LinkType.Folder, folderId, BatchUpdateAssetLinks(ids)).updatedAssetIds.size)
        assertEquals(2, assetService.removeLinks(LinkType.Folder, folderId, ids).updatedAssetIds.size)
        assertEquals(0, assetService.removeLinks(LinkType.Folder, folderId, ids).updatedAssetIds.size)
    }

    @Test
    fun testSetPermissions() {
        val perm = permissionService.getPermission(Groups.ADMIN)
        val result = assetService.setPermissions(
                BatchUpdatePermissionsRequest(AssetSearch(), Acl().addEntry(perm.id, 4)))

        val page = searchService.search(Pager.first(), AssetSearch())
        for (asset in page) {
            val perms = asset.getAttr("system.permissions", PermissionSchema::class.java)
            assertTrue(perms!!.export.contains(perm.id))
        }
    }

    @Test
    fun testSetPermissionsWithReplace() {
        val perm = permissionService.getPermission(Groups.ADMIN)
        val result = assetService.setPermissions(
                BatchUpdatePermissionsRequest(AssetSearch(), Acl().addEntry(perm.id, 4), replace = true))

        val page = searchService.search(Pager.first(), AssetSearch())
        for (asset in page) {
            val perms = asset.getAttr("system.permissions", PermissionSchema::class.java)
            assertTrue(perms!!.export.contains(perm.id))
            assertTrue(perms!!.read.isEmpty())
            assertTrue(perms!!.write.isEmpty())
        }
    }

    @Test
    fun testCreateFieldEdit() {
        val title = "khaaaaaan"
        val field = fieldSystemService.getField("media.title")

        val page = searchService.search(Pager.first(), AssetSearch())
        for (asset in page) {
            assetService.createFieldEdit(
                    FieldEditSpec(UUID.fromString(asset.id), field.id, null,title))
        }
        for (asset in searchService.search(Pager.first(), AssetSearch()).list) {
            assertEquals(title, asset.getAttr("media.title", String::class.java))
        }

        val result =
                auditLogDao.getAll(AuditLogFilter(
                        fieldIds=listOf(field.id),
                        types=listOf(AuditLogType.Changed)))
        assertTrue(result.size() > 0)
    }

    @Test
    fun testDeleteFieldEdit() {
        val title = "khaaaaaan"
        val field = fieldSystemService.getField("media.title")

        var page = searchService.search(Pager.first(), AssetSearch())
        var asset = page.list[0]
        val edit = assetService.createFieldEdit(
                FieldEditSpec(UUID.fromString(asset.id), field.id, null, title))
        assertTrue(assetService.deleteFieldEdit(edit))

        page = searchService.search(Pager.first(), AssetSearch())
        for (asset in page) {
            assertNotEquals(title, asset.getAttr("media.title", String::class.java))
        }

        val result =
                auditLogDao.getAll(AuditLogFilter(
                        fieldIds=listOf(field.id),
                        types=listOf(AuditLogType.Changed)))

        assertTrue(result[0].message!!.contains("undo"))
        assertTrue(result[1].message!!.contains("manual edit"))

    }

    /**
     * Just checks we didn't lose any system attributes after
     * reprocessing, does not validate any values.
     */
    fun validateSystemAttrsExist(asset: Document) {
        assertNotNull(asset.getAttr("system.permissions"))
        assertNotNull(asset.getAttr("system.timeCreated"))
        assertNotNull(asset.getAttr("system.timeModified"))
        assertNotNull(asset.getAttr("system.links"))
    }
}